source-dropbox/main.go

```diff
@@ -0,0 +1,272 @@
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	token, err := cfg.Credentials.GetAccessToken(ctx)
+	if err != nil {
+		return &dropboxStore{}, err
+	}
+	config := dropbox.Config{
+		Token:    token,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks the connection to the Azure Blob Storage container.
+// It returns an error if the container is not found, the account is disabled, or if there is an authorization failure.
+// If the listing is successful, it returns nil.
+func (dbx *dropboxStore) check() error {
+	_, err := dbx.client.ListFolder(&files.ListFolderArg{Path: dbx.config.Path})
+	if err != nil {
+		return fmt.Errorf("failed to list files: %w", err)
+	}
+	return nil
+}
+
+func (dbx *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	log.Debug("Listing files in path: ", dbx.config.Path)
+	log.Debug("Query: ", query)
+	files, err := dbx.client.ListFolder(&files.ListFolderArg{Path: query.Prefix})
+	if err != nil {
+		return nil, err
+	}
+
+	return &dropboxListing{
+		ctx:    ctx,
+		client: dbx.client,
+		files:  *files,
+		index:  0,
+	}, nil
+}
+
+func (dbx *dropboxStore) Read(ctx context.Context, obj filesource.ObjectInfo) (io.ReadCloser, filesource.ObjectInfo, error) {
+	log.Debug("Reading object: ", obj.Path)
+	meta, reader, err := dbx.client.Download(&files.DownloadArg{Path: obj.Path})
+	if err != nil {
+		return nil, filesource.ObjectInfo{}, err
+	}
+
+	obj.Path = meta.PathDisplay
+	obj.ModTime = meta.ClientModified
+	if meta.ExportInfo != nil {
+		obj.ContentType = meta.ExportInfo.ExportAs
+	}
+	obj.Size = int64(meta.Size)
+
+	return reader, obj, nil
+}
+
+type dropboxListing struct {
+	ctx    context.Context
+	client files.Client
+	files  files.ListFolderResult
+	index  int
+}
+
+func (l *dropboxListing) Next() (filesource.ObjectInfo, error) {
+	log.Debug("Listing next file")
+	log.Debug(l.files.Entries)
+
+	if len(l.files.Entries) == 0 {
+		return filesource.ObjectInfo{}, io.EOF
+	}
+	if !l.files.HasMore && l.index > 0 {
+		log.Debug("No more files to list")
+		return filesource.ObjectInfo{}, io.EOF
+	}
+	entry, ok := l.files.Entries[l.index].(*files.FileMetadata)
```

Comment:
We need to handle subdirectories. When I tested this configured with a folder containing a subdirectory it crashed. There's also a "deleted file" metadata type that we may need to handle. We will need to recursively list the contents of the subdirectory I'd imagine. I think we do this in `source-s3` etc. by returning an object with `IsPrefix: true` for directories, see https://github.com/estuary/connectors/blob/2c4ad1bcedde33800518f7125d9ef94f4e498c72/source-s3/main.go#L214-L217